### ISSUE TITLE
Module "includes" not equivalent to rebinding for types using first-class modules

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#5749]
**Original bug ID:** 5749
**Reporter:** @mmottl
**Status:** acknowledged (set by @damiendoligez on 2012-09-27T12:33:23Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 4.00.0
**Category:** typing
**Monitored by:** @gasche mehdi "Julien Signoles" @hcarty @mmottl

## Bug description

The following code fails with a signature mismatch:

```ocaml
module type Spec = sig
  module type M = sig end
  type t = (module M) list
end

module Make (Spec : Spec) : sig type t = Spec.t end = struct
  include Spec
  (* type t = Spec.t *)
end
```

But commenting out "include Spec" while uncommenting the succeeding line containing the rebinding works without problems.

The compiler presumably treats the module type bound in the functor body as different from the one in the Spec-module.  Since the type of first-class modules has to be the same nominally, not just structurally, this might explain the observed error message.  But here the equivalence really goes beyond structure, the "include" guarantees that both module types refer to the same thing.  I'm not sure, but maybe this situation can be improved?

**Comment by vicuna** [comment]
**Comment author:** @alainfrisch

This seems hard to fix.  Within the functor's body, the include statement basically amounts to inlining:

```ocaml
  module type M = sig end
  type t = (module M) list
```

which produces a different module type name M.  We would like to keep both equations "type t = (module M) list" and "type t = Spec.t", but this is not possible in the current type system.

I don't see a simple solution.  At some point, it would be interesting to explore relaxing the nominal comparison of module types in package types, but I'm sure it will raise a lot of other tricky issues.

Did you find this problematic case in "real code"?

**Comment by vicuna** [comment]
**Comment author:** @mmottl

Oh well, I guess we will have to live with that until the day that some superhero rewrites the type system from scratch.

Yes, I ran into this in "real code" (experimenting with ways to represent operators on covariance functions), but it's still easy enough to work around this problem.

**Comment by yallop** [comment]
This is fixed, possibly by #6333, and the example is now accepted as written.
